aehmc/proposals.py

```diff
@@ -0,0 +1,159 @@
+
+    Choose between the current proposal and the proposal built from the last
+    trajectory state.
+
+    Parameters
+    ----------
+    srng
+        RandomStream object
+    proposal
+        The current proposal, it does not necessarily correspond to the
+        previous state on the trajectory
+    new_proposal
+        The proposal built from the last trajectory state.
+
+    Return
+    ------
+    Either the current or the new proposal.
+
+    """
+    state, energy, weight, sum_log_p_accept = proposal
+    new_state, new_energy, new_weight, new_sum_log_p_accept = new_proposal
+
+    p_accept = aet.expit(new_weight - weight)
+    do_accept = srng.bernoulli(p_accept)
+    updated_proposal = maybe_update_proposal(do_accept, proposal, new_proposal)
+
+    return updated_proposal
+
+
+def progressive_biased_sampling(
+    srng: RandomStream, proposal: ProposalStateType, new_proposal: ProposalStateType
+) -> ProposalStateType:
+    """Baised proposal sampling.
+
+    Choose between the current proposal and the proposal built from the last
+    trajectory state. Unlike uniform sampling, biased sampling favors new
+    proposals. It thus biases the transition away from the trajectory's initial
+    state.
+
+    Parameters
+    ----------
+    srng
+        RandomStream object
+    proposal
+        The current proposal, it does not necessarily correspond to the
+        previous state on the trajectory
+    new_proposal
+        The proposal built from the last trajectory state.
+
+    Return
+    ------
+    Either the current or the new proposal.
+
+    """
+    state, energy, weight, sum_log_p_accept = proposal
+    new_state, new_energy, new_weight, new_sum_log_p_accept = new_proposal
+
+    p_accept = aet.clip(aet.exp(new_weight - weight), 0.0, 1.0)
+    do_accept = srng.bernoulli(p_accept)
+    updated_proposal = maybe_update_proposal(do_accept, proposal, new_proposal)
+
+    return updated_proposal
+
+
+def maybe_update_proposal(
+    do_accept: bool, proposal: ProposalStateType, new_proposal: ProposalStateType
+) -> ProposalStateType:
+    """Return either proposal depending on the boolean `do_accept`"""
+    state, energy, weight, sum_log_p_accept = proposal
+    new_state, new_energy, new_weight, new_sum_log_p_accept = new_proposal
+
+    updated_weight = _logaddexp(weight, new_weight)
+    updated_sum_log_p_accept = _logaddexp(sum_log_p_accept, new_sum_log_p_accept)
+
+    updated_q = aet.where(do_accept, new_state[0], state[0])
+    updated_p = aet.where(do_accept, new_state[1], state[1])
+    updated_potential_energy = aet.where(do_accept, new_state[2], state[2])
+    updated_potential_energy_grad = aet.where(do_accept, new_state[3], state[3])
+    updated_energy = aet.where(do_accept, new_energy, energy)
+
+    return (
+        (updated_q, updated_p, updated_potential_energy, updated_potential_energy_grad),
+        updated_energy,
+        updated_weight,
+        updated_sum_log_p_accept,
+    )
+
+
+def _logaddexp(a, b):
```

Comment:
This was implemented twice. 

Also a reminder for me to check how this one performs compared with the more general default `logaddexp` rewrite we have. We can easily add this when there are only two terms being added, if it's any better

Reply:
Yes we should definitely check against the rewrite. In the meantime I am not so bothered by the duplication.